daffodil-runtime1/src/main/scala/org/apache/daffodil/runtime1/processors/VariableMap1.scala

```diff
@@ -311,31 +306,33 @@ class VariableMap private (vTable: Map[GlobalQName, ArrayBuffer[VariableInstance
    * expression
    */
   def setFirstInstanceInitialValues(): Unit = {
-    vTable.foreach {
-      case (_, variableInstances) => {
-        variableInstances(0).firstInstanceInitialValue = variableInstances(0).value
-      }
+    vTable.foreach { variableInstances =>
+      variableInstances.head.firstInstanceInitialValue = variableInstances.head.value
     }
   }
 
+  /**
+   * Performance of this is linear in number of variables, this should not be used in
+   * performance critical sections.
+   */
   def find(qName: GlobalQName): Option[VariableInstance] = {
-    val optBuf = vTable.get(qName)
-    val variab = {
-      if (optBuf.isDefined)
-        Some(optBuf.get.last)
-      else
-        None
-    }
-    variab
+    getVariableRuntimeData(qName).map { vrd => vTable(vrd.vmapIndex).head }
```

Comment:
Isn't a hash map of qname -> vmapIndex just a constant table that can be computed once we have all the variable declarations? So this can be constant time.  Your representation avoids having to call this much as a VRD knows its index, so getting from VRD to variable on top of stack is constant time, but I think this find can also  be constant time. 

This would only be important if there were more than a trivial number of variables. Linear search may in fact be the fastest thing. I think most cases the number of variables is < 6.

Reply:
Yeah, I considered doing that, but the only places we do a qname lookup are in layer compilation (if they use variables), CLI debugging, or setting external variables. Only that last one *really* matters for performance, but 1) setting external variables is fairly rare and 2) setExternalVariables has more compliated logic than a hash lookup--it considers local name, detects ambiguities if there's no namspace, etc. As you point out linear search is probably fine for these cases with the usual few number of variables.

In reality, these functions probably aren't that slow, but I added the comment because I don't want us to accidentally use them in actual parsing/unparsing logic or something where speed really is critical.